activerecord/lib/active_record/associations/association_scope.rb

```diff
@@ -159,6 +159,14 @@ def add_constraints(scope, owner, refl, chain_head, chain_tail)
           scope
         end
 
+        def apply_scope(scope, table, key, value)
+          if scope.table == table
+            scope.where!(key => value)
```

Comment:
The `AssociationScope` returns newly created `scope` from `#scope`.
`#scope` creates `scope = klass.unscoped` and mutates the `scope` by `scope.extending! reflection.extensions`.

```ruby
      def scope(association)
        klass = association.klass
        reflection = association.reflection
        scope = klass.unscoped
        owner = association.owner
        alias_tracker = AliasTracker.create(klass.connection, klass.table_name)
        chain_head, chain_tail = get_chain(reflection, association, alias_tracker)

        scope.extending! reflection.extensions # mutate the `scope`
        add_constraints(scope, owner, reflection, chain_head, chain_tail)
      end
```

`#add_constraints` also mutates the `scope` (I have another bug fix PR #29659. it also mutate the `scope` in `#add_constraints`).

```ruby
      private

        def add_constraints(scope, owner, refl, chain_head, chain_tail)
          owner_reflection = chain_tail
          table = owner_reflection.alias_name
          scope = last_chain_scope(scope, table, owner_reflection, owner)

          reflection = chain_head
          while reflection
            table = reflection.alias_name
            next_reflection = reflection.next

            unless reflection == chain_tail
              foreign_table = next_reflection.alias_name
              scope = next_chain_scope(scope, table, reflection, foreign_table, next_reflection)
            end

            reflection.constraints.each do |scope_chain_item|
              item = eval_scope(reflection, table, scope_chain_item, owner)

              if scope_chain_item == refl.scope
                scope.merge! item.except(:where, :includes) # mutate the `scope`
              end

              reflection.all_includes do
                scope.includes! item.includes_values # mutate the `scope`
              end

              scope.unscope!(*item.unscope_values) # mutate the `scope`
              scope.where_clause += item.where_clause # mutate the `scope`
              scope.order_values |= item.order_values # mutate the `scope`
            end

            reflection = next_reflection
          end

          scope
        end
```

`#last_chain_scope` and `#next_chain_scope` are called via `#add_constraints`.
Since `#add_constraints` mutates the `scope`, I think it is no worth to avoid mutations in `#last_chain_scope` and `#next_chain_scope`. In addition, extra `spawn` to avoid mutations has a overhead (ref https://github.com/rails/rails/pull/29009#issuecomment-321773151).

I placed the `apply_scope` at the bottom of private methods because the common logic is only used for internal private methods. I think that usually users don't read the code from the bottom of private methods.

WDYT?